### ISSUE TITLE
Script is triggered by a datapoint: JS stops. 

### ISSUE THREAD

**Issue by MichaelSchaaf999** [ioBroker/ioBroker.javascript#1234]
**Describe the bug**  
I added the log entries below.
I am not experienced here, but what I can point to is a line with: "JavaScript heap out of memory".
But the iOBroker in the Proxmox VM has enough memory as shown here:
![image](https://user-images.githubusercontent.com/53481090/212757730-e737a651-8f6d-47b8-8a26-86f7bd66454e.png)

This script is new, never ran before.

**To Reproduce**  
I can reproduce this abort by activating the script and then open a door (that is monitored by this script).

**Expected behavior**  
No abort please.

**Screenshots & Logfiles**  
not applicable

**Versions:**  
 - Adapter version: Javascript.0 v6.1.4
 - JS-Controller version: <js-controller-version>  4.0.24
 - Node version: <node-version>  v14.20.0
 - Operating system: <os-name>
![image](https://user-images.githubusercontent.com/53481090/212758249-360bf09c-64c7-42b2-b49f-1da25f9bdb54.png)
in a Proxmox VM

**Additional context**  
not applicable

Best regards
Michael

javascript.0 | 2023-01-16 20:43:36.958 | info | received all objects
-- | -- | -- | --
javascript.0 | 2023-01-16 20:43:36.903 | info | received all states
javascript.0 | 2023-01-16 20:43:35.818 | info | requesting all objects
javascript.0 | 2023-01-16 20:43:35.817 | info | requesting all states
javascript.0 | 2023-01-16 20:43:35.074 | info | starting. Version 6.1.4 in /opt/iobroker/node_modules/iobroker.javascript, node: v14.20.0, js-controller: 4.0.24
host.iobroker | 2023-01-16 20:43:03.974 | warn | instance system.adapter.javascript.0 terminated due to SIGABRT
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[24]: 24: 0x1448f59 [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[23]: 23: 0x10b5dae v8::internal::Runtime_ObjectEntries(int, unsigned long*, v8::internal::Isolate*) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[22]: 22: 0xf2dc18 v8::internal::GetOwnValuesOrEntries(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::PropertyFilter, bool, bool) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[21]: 21: 0xf2d9b4 v8::internal::FastGetOwnValuesOrEntries(v8::internal::Isolate*, v8::internal::Handle<v8::internal::JSReceiver>, bool, v8::internal::Handle<v8::internal::FixedArray>*) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[20]: 20: 0xd4ab94 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::internal::AllocationType) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[19]: 19: 0xd4aad9 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawArray(int, v8::internal::AllocationType) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[18]: 18: 0xd4e87d v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[17]: 17: 0xd890ac v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[16]: 16: 0xd85c68 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[15]: 15: 0xd84178 v8::internal::Heap::MarkCompact() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.974 | error | Caught by controller[14]: 14: 0xdc7eb1 v8::internal::MarkCompactCollector::CollectGarbage() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[13]: 13: 0xdb5eb5 v8::internal::MarkCompactCollector::Evacuate() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[12]: 12: 0xdb5cec v8::internal::MarkCompactCollector::EvacuatePagesInParallel() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[11]: 11: 0xdb5450 void v8::internal::MarkCompactCollectorBase::CreateAndExecuteEvacuationTasks<v8::internal::FullEvacuator, v8::internal::MarkCompactCollector>(v8::internal::MarkCompactCollector*, v8::internal::ItemParallelJob*, v8::internal::MigrationObserver*, long) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[10]: 10: 0xd921e9 v8::internal::ItemParallelJob::Run() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[9]: 9: 0xd9f908 v8::internal::PageEvacuationTask::RunInParallel(v8::internal::ItemParallelJob::Task::Runner) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[8]: 8: 0xd9f68f v8::internal::Evacuator::EvacuatePage(v8::internal::MemoryChunk*) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[7]: 7: 0xdb34f6 v8::internal::FullEvacuator::RawEvacuatePage(v8::internal::MemoryChunk*, long*) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[6]: 6: 0xda74be v8::internal::EvacuateNewSpaceVisitor::Visit(v8::internal::HeapObject, int) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[5]: 5: 0xd76b25 [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[4]: 4: 0xbba907 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[3]: 3: 0xbba58e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[2]: 2: 0x970199 node::FatalError(char const*, char const*) [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[1]: 1: 0xa3ac10 node::Abort() [io.javascript.0]
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[0]: FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
host.iobroker | 2023-01-16 20:43:03.973 | error | Caught by controller[0]: <--- JS stacktrace --->
host.iobroker | 2023-01-16 20:43:03.972 | error | Caught by controller[0]: [400:0x486edb0] 567937 ms: Mark-sweep (reduce) 2047.6 (2055.2) -> 2047.3 (2055.9) MB, 2707.3 / 0.0 ms (+ 0.0 ms in 16 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 2710 ms) (average mu = 0.073, current mu = 0.001) a[400:0x486edb0] 569496 ms: Mark-sweep (reduce) 2048.3 (2052.9) -> 2048.1 (2054.7) MB, 1502.4 / 0.0 ms (+ 53.5 ms in 16 steps since start of marking, biggest step 6.4 ms, walltime since start of marking 1560 ms) (average mu = 0.043, current mu = 0.002)
host.iobroker | 2023-01-16 20:43:03.971 | error | Caught by controller[0]: <--- Last few GCs --->

**Comment by Apollon77** [comment]
Heap !== RAM ... in fact RAM seems to go over 2GB, so it seems that one of your scripts is having a memory leak ...

Options: Updare Node.js to current 16 to see if it is fixed there

If it is not caused by node.js version iself then you need to find the script ... One idea could e to split your scripts onmultiple javascript instances and see which instance is then collecting RAM and then move scripts one by one or such. You can monitor the RAM states under system.adapter.javascript.X.mem* to see earlier if it happens again.

**Comment by MichaelSchaaf999** [comment]
These two values you are talking about?
-> system.adapter.javascript.X.mem*

![image](https://user-images.githubusercontent.com/53481090/212821225-cc97b992-1c2c-466c-9503-b95ff19f2e62.png)

And: I already know the script, it's the last new one. Immediately causes this, what you call: memory leak. 
I'll have to inspect and somehow debug that script.

**Comment by MichaelSchaaf999** [comment]
Found it.
Bad coding.
A repeat loop for a function where multiple speak commands each with a delay of 15secs where thought to be executed.
But I assume that the repeat loop just produced code or RAM usage for this and the repeat loop was faster than the delayed speak commands: So no chance to end that function.

Thanks to your support I was able to identify the script and then the bad coding.
Best regards
M
